pandas/core/categorical.py

```diff
@@ -294,16 +293,8 @@ def __init__(self, values, categories=None, ordered=None, dtype=None,
 
         # sanitize input
         if is_categorical_dtype(values):
-
-            # we are either a Series or a CategoricalIndex
-            if isinstance(values, (ABCSeries, ABCCategoricalIndex)):
-                values = values._values
-
-            if ordered is None:
-                ordered = values.ordered
-            if categories is None:
-                categories = values.categories
-            values = values.get_values()
+            if dtype.categories is None:
+                dtype = CategoricalDtype(values.categories, dtype.ordered)
```

Comment:
some lines above, there is already a 

```
         elif is_categorical(values):
             dtype = values.dtype._from_categorical_dtype(values.dtype,
                                                          categories, ordered)
```

which should catch the same ?

Reply:
Yes... but instead testing ``dtype.categories`` must be done here (that is, _after_ all the branches of the above ``if...elif...else``). Viceversa, the ``dtype`` must be defined _before_ the ``fastpath`` test.

Reply:
OK, but help me understand in what case this line is needed and is doing something different as the already defined `dtype`

Reply:
For instance the case in which ``values`` is already a ``Categorical``, ``categories`` is ``None`` and ``dtype`` is ``"category"``.

Reply:
and in that case 

```
         elif is_categorical(values):
             dtype = values.dtype._from_categorical_dtype(values.dtype,
                                                          categories, ordered)
```

will already have created an appropriate `dtype` ? 
So I still don't understand why in such a case it would need to be re-determined.

Reply:
> will already have created an appropriate dtype ?

No, precisely because ``categories`` is ``None``. Not just in theory: you can actually try and verify that tests break.
It _could_ be possible to first have a different set of tests which finds the right value for ``categories``, but I doubt it would result in simpler code.

Reply:
Assume this small example:

```
In [1]: values = pd.Categorical(['a', 'b', 'c', 'a'])

In [2]: values
Out[2]: 
[a, b, c, a]
Categories (3, object): [a, b, c]

In [4]: categories = None

In [5]: dtype = 'category'

In [9]: ordered = None

In [10]: dtype = values.dtype._from_categorical_dtype(values.dtype, categories, ordered)

In [11]: dtype
Out[11]: CategoricalDtype(categories=['a', 'b', 'c'], ordered=False)
```

So if you pass categorical values, it will by definition have categories (can a Categorical have categories of None?), and those will be passed to the resulting `dtype` object returned from `values.dtype._from_categorical_dtype`. 
So still wondering, in what case can you have categorical values, but where the resulting `dtype` from above still has `dtype.categories is None`?

Reply:
So when removing that check, it is the case where a `CategoricalDType` instance is passed that differs from the dtype of the values:

```
dtype = CategoricalDtype(None, ordered=True)
values = Categorical(['a', 'b', 'd'])
Categorical(values, dtype=dtype)
```

I would find that more logical to handle in the `if dtype is not None:` path (to clearly see which takes precedence in such a case). But I assume that when `fastpath=True` we don't want to check that. Although I think that if you use that, dtype.categories will never be None.